Fixed cond with group and summaries, issue #24815

ila96 opened the pull request.

alextp reviewed (changes requested):
Can you write a test which fails before this PR and passes after?

I'd like to prevent this issue from regressing in future changes to TF.

ila96 committed:
Added test for cond with grouped summaries in one branch

Added test function testCondWithGroupAndSummaries into
control_flow_ops_test.py, that checks whether graph is created
succesfully when one branch of cond contains grouped summaries.

Signed-off-by: Nemanja Ilic <ila.96@live.com>

ila96 commented:
> Can you write a test which fails before this PR and passes after?
> 
> I'd like to prevent this issue from regressing in future changes to TF.

Thank You @alextp, I added test You asked for.

alextp commented on tensorflow/python/ops/control_flow_ops_test.py:
```diff
@@ -417,6 +419,16 @@ def testCondDuplicateArg2(self):
+              random_ops.random_normal([]) >= 0,
+              lambda: control_flow_ops.group(summary.scalar('loss', 0.2)),
+              control_flow_ops.no_op)
+    except TypeError as e:
```
I'd rather you have the test fail only with a statement of the form self.assert*, like self.assertAllEqual, self.assertRaises, etc.

In this one I'd validate some property of the op, like its number or outputs or sometihng like that.

ila96 replied:
If I was to do something like that test would fail without this fix, exception TypeError would be raised on creating branch (it wouldn't even come to the assert that is validating op properties) and that's why I introduced try block so the test would fail in case that exception is raised. Is it better to do it with self.assertEqual?

alextp replied:
I'd rather write this test as something like

```
     writer = summary.create_file_writer('tb')
      with writer.as_default(), summary.always_record_summaries():
          op = control_flow_ops.cond(
              random_ops.random_uniform([]) >= -1,  # note: will always be true
              lambda: control_flow_ops.group(summary.scalar('loss', 0.2)),
              control_flow_ops.no_op)
      self.assertEqual(len(op.outputs), 0)
```

This way there's an assertion and if the error is raised the user will get a failure anyway

ila96 replied:
I modified in the way you suggested, please check it out.

ila96 commented:
I updated files, added whitespaces and removed unnecessary import, because //tensorflow/tools/ci_build:gen_ci_sanity_out failed.

rthadur commented:
@ila96 can you please check ubuntu sanity errors ?

ila96 commented:
> @ila96 can you please check ubuntu sanity errors ?

Hi, I run tensorflow/tools/ci_build/ci_sanity.sh and got errors that don't contain the files I changed. Also, ubuntu sanity error log doesn't list files I changed. Here's a snippet of a log:
"tensorflow/contrib/gan/python/train.py:543: [C0301(line-too-long), ] Line too long (94/80)"
I'm not sure what to do?

gbaned commented:
@ila96 please resolve conflicts.

gbaned commented:
@ila96 gentle ping to resolve conflicts. Thanks!

ila96 committed:
Fixed cond with grouped summaries in one branch

This commit fixes graph creation error when condition branch contains
summaries. Fixed by not performing Identity op in case when
original_result has type Operation.

This is fix for issue #24815.

The fix is covered by test testCondWithGroupAndSummaries in
control_flow_ops_test.py.

Signed-off-by: Nemanja Ilic <ila.96@live.com>

ila96 commented:
I rebased and solved conflicts, and run ci_sanity.sh but it didn't show files i changed.

gbaned commented:
@ila96 Can you please resolve the conflicts? Thanks!

ila96 commented:
Hi @alextp, @gbaned, can You please merge this pull request?

alextp commented on tensorflow/python/ops/control_flow_ops_test.py:
```diff
@@ -22,6 +22,7 @@
 import numpy as np
 
 from tensorflow.python import tf2
+from tensorflow.contrib import summary
```
TF core tests are not allowed to depend on contrib

ila96 replied:
Ok, but this is fix for that problem, as mentioned in issue #24815 . Do You have any advice on how to rewrite test?

alextp replied:
Import the internal module, which I think is in tensorflow.python.ops.summary_v2

ila96 commented:
Hi @alextp, @gbaned, please feel free to let me know if changes are needed.

mellanox-github commented:
Can one of the admins verify this patch?

rthadur commented on tensorflow/python/ops/control_flow_ops_test.py:
```diff
@@ -441,6 +442,16 @@ def testCond_gradient(self):
                      self.evaluate(grads_false[0]))
 
+  def testCondWithGroupAndSummaries(self):
+    writer = summary_ops_v2.create_file_writer('tb')
```
this test case is failing due to this error : `PermissionDeniedError: Localfile::CreateDir mkdir failed on tb: Permission denied` , can you please check this error

tensorflowbutler commented:
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

gbaned commented:
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!
